ducktape/cluster/vagrant.py

```diff
@@ -22,12 +26,49 @@ class VagrantCluster(JsonCluster):
     """
     An implementation of Cluster that uses a set of VMs created by Vagrant. Because we need hostnames that can be
     advertised, this assumes that the Vagrant VM's name is a routeable hostname on all the hosts.
+
+    - If cluster_file is specified in the constructor's kwargs (i.e. passed via command line argument --cluster-file)
+      - If cluster_file exists on the filesystem, read cluster info from the file
+      - Otherwise, retrieve cluster info via "vagrant ssh-config" from vagrant and write cluster info to cluster_file
+    - Otherwise, retrieve cluster info via "vagrant ssh-config" from vagrant
     """
 
-    def __init__(self):
-        hostname, ssh_hostname, username, flags = None, None, None, ""
-        nodes = []
+    def __init__(self, *args, **kwargs):
+        is_read_from_file = False
+
+        cluster_file = kwargs.get("cluster_file")
+        if cluster_file is not None:
+            try:
+                cluster_json = json.load(open(os.path.abspath(cluster_file)))
+                is_read_from_file = True
+            except IOError:
+                # It is OK if file is not found. Call vagrant ssh-info to read the cluster info.
+                pass
+
+        if not is_read_from_file:
+            cluster_json = {
+                "nodes": self._get_nodes_from_vagrant()
+            }
+
+        super(VagrantCluster, self).__init__(cluster_json)
+
+        is_aws = self._is_aws()
+        if not is_read_from_file:
+            # go through and find fully qualified domain name for each node
+            # this makes it possible to not require write access to /etc/hosts on the test driver machine
+            externally_routable_ip = {}
```

Comment:
Suppose the cluster info is read from file, then `JsonCluster` will be initialized with the `externally_routable_ip`.

Otherwise, it is true that `JsonCluster` will be initialized without `externally_routable_ip`. However, this seems inevitable. This is because `RemoteAccount` needs to be created in `JsonCluster`, `externally_routable_ip` for vagrant cluster should be read in `VagrantCluster` via ssh command, and the ssh command should be issued via `RemoteAccount`. I don't have a better solution here. Do you have a solution for this?